src/app/modules/webhook/webhook.consumer.ts

```diff
@@ -0,0 +1,247 @@
+import { WebhookProducer } from './webhook.producer'
+import * as WebhookService from './webhook.service'
+import { isSuccessfulResponse } from './webhook.utils'
+
+const logger = createLoggerWithLabel(module)
+const EncryptSubmission = getEncryptSubmissionModel(mongoose)
+
+/**
+ * Starts polling a queue for webhook messages.
+ * @param queueUrl URL of queue from which to consume messages
+ * @param producer Producer which can be used to enqueue messages
+ */
+export const startWebhookConsumer = (
+  queueUrl: string,
+  producer: WebhookProducer,
+): void => {
+  const app = Consumer.create({
+    queueUrl,
+    region: config.aws.region,
+    handleMessage: createWebhookQueueHandler(producer),
+    // By default, the default Node.js HTTP/HTTPS SQS agent
+    // creates a new TCP connection for every new request.
+    // In production, pass an SQS instance to avoid the cost
+    // of establishing new connections.
+    sqs: config.isDev
+      ? undefined
+      : new aws.SQS({
+          region: config.aws.region,
+          httpOptions: {
+            agent: new https.Agent({
+              keepAlive: true,
+            }),
+          },
+        }),
+  })
+
+  app.on('error', (error, message) => {
+    logger.error({
+      message:
+        'Webhook consumer encountered error while interacting with queue',
+      meta: {
+        action: 'startWebhookConsumer',
+        message,
+      },
+      error,
+    })
+  })
+
+  app.start()
+
+  logger.info({
+    message: 'Webhook consumer started',
+    meta: {
+      action: 'startWebhookConsumer',
+    },
+  })
+}
+
+/**
+ * Creates a handler to consume messages from webhook queue.
+ * This handler does the following:
+ * 1) Parses the message
+ * 2) If the webhook is not due, requeues the message
+ * 3) If the webhook is due, attempts the webhook
+ * 4) Records the webhook attempt in the database
+ * 5) If the webhook failed again, requeues the message
+ * @param producer Producer which can write messages to queue
+ * @returns Handler for consumption of queue messages
+ */
+const createWebhookQueueHandler = (producer: WebhookProducer) => async (
+  sqsMessage: aws.SQS.Message,
+): Promise<void> => {
+  logger.info({
+    message: 'Consumed message from webhook queue',
+    meta: {
+      action: 'createWebhookQueueHandler',
+    },
+  })
+  const { Body } = sqsMessage
+  if (!Body) {
+    logger.error({
+      message: 'Webhook queue message contained undefined body',
+      meta: {
+        action: 'createWebhookQueueHandler',
+        sqsMessage,
+      },
+    })
+    // Malformed message will be retried until redrive policy is exceeded,
+    // upon which it will be moved to dead-letter queue
+    return Promise.reject()
+  }
+
+  // Parse message
+  const webhookMessageResult = WebhookQueueMessage.deserialise(Body)
+  if (webhookMessageResult.isErr()) {
+    logger.error({
+      message: 'Webhook queue message could not be parsed',
```

Comment:
SQS should provide a globally unique identifier for every message. Can we include that for every log message, so that there's no ambiguity if we need to track down a message?